autoscheduler/frontend/src/tests/ui/SchedulePreview.test.tsx

```diff
@@ -145,4 +147,94 @@ describe('SchedulePreview component', () => {
       });
     });
   });
+
+  describe('saves the correct schedule', () => {
+    test('when the first schedule is saved', async () => {
+      // arrange
+      const store = createStore(autoSchedulerReducer);
+      const { findAllByText } = render(
+        <Provider store={store}>
+          <SchedulePreview />
+        </Provider>,
+      );
+      store.dispatch(replaceSchedules([testSchedule1, testSchedule2]));
+
+      // act
+      const saveScheduleButton = (await findAllByText('Save'))[0];
+      fireEvent.click(saveScheduleButton);
+
+      // assert
+      await waitFor(() => (
+        expect(store.getState().schedules.savedSchedules).toContainEqual(testSchedule1)
+      ));
+    });
+
+    test('when a schedule with index greater than 0 is saved', async () => {
+      // arrange
+      const store = createStore(autoSchedulerReducer);
+      const { findAllByText } = render(
+        <Provider store={store}>
+          <SchedulePreview />
+        </Provider>,
+      );
+      store.dispatch(replaceSchedules([testSchedule1, testSchedule2]));
+
+      // act
+      const saveScheduleButton = (await findAllByText('Save'))[1];
+      fireEvent.click(saveScheduleButton);
+
+      // assert
+      await waitFor(() => (
+        expect(store.getState().schedules.savedSchedules).toContainEqual(testSchedule2)
+      ));
+    });
+  });
+
+  describe('unsaves the correct schedule', () => {
+    test('when the first schedule is unsaved', async () => {
+      // arrange
+      const store = createStore(autoSchedulerReducer);
+      const { findAllByText, getByText } = render(
+        <Provider store={store}>
+          <SchedulePreview />
+        </Provider>,
+      );
+      store.dispatch(replaceSchedules([testSchedule1, testSchedule2]));
+
+      // act
+      // save, wait for save to finish, then unsave
+      const saveScheduleButton = (await findAllByText('Save'))[0];
+      fireEvent.click(saveScheduleButton);
+      await waitFor(() => getByText('Unsave'));
+      fireEvent.click(saveScheduleButton);
+
+      // assert
+      await waitFor(() => (
+        expect(store.getState().schedules.savedSchedules).toHaveLength(0)
+      ));
+    });
+
+    test('when a schedule with index greater than 0 is unsaved', async () => {
+      // arrange
+      const store = createStore(autoSchedulerReducer);
+      const { findAllByText, getByText } = render(
+        <Provider store={store}>
+          <SchedulePreview />
+        </Provider>,
+      );
+      store.dispatch(replaceSchedules([testSchedule1, testSchedule2]));
+
+      // act
+      // save, wait for save to finish, then unsave
+      const saveScheduleButton = (await findAllByText('Save'))[1];
+      fireEvent.click(saveScheduleButton);
+      await waitFor(() => getByText('Unsave'));
+      fireEvent.click(saveScheduleButton);
+
+      // assert
+      await waitFor(() => (
+        expect(store.getState().schedules.savedSchedules).toHaveLength(0)
+      ));
+    });
```

Comment:
These tests cause the following error:
![image](https://user-images.githubusercontent.com/28655462/84580399-3f7ecf00-ad9c-11ea-8b94-a6b1b83259ec.png)
This is because of the line `await waitFor(() => getByText('Unsave'));` (which I realize could just be `await findByText('Unsave')`, not sure why I didn't do that in hindsight but it still results in the same error). This error doesn't occur when actually using the UI so it's pretty minor and the tests still work correctly despite it but still, it would be nice to fix.

Reply:
I would move the `sleep` function to `testUtil.ts` and mock it to return immediately during the test

Reply:
I just removed the sleep function since everyone's seen how it looks and we won't want it in master anyways